Desktop notification windows

ajCameron opened the pull request:
Tested and working on ubuntu 20.04 LTS, WLS and windows.

Some patching required to make mewbot run happily on windows - in particular, can't install the signal handlers.

setup.py modified to install win10toast if on windows.

ajCameron committed:
Need to test again before push, and discuss a) monkeypatching b) system dependant requirements

ajCameron committed:
Example works, some tidying done, linting done.

ajCameron committed:
blacking the examples would have helped...

ajCameron committed:
Running the linters on the right branch does, in fact, help

ajCameron committed:
More linting fixes

ajCameron commented:
Something seems to be going really badly wrong with loading more than one IOConfig.

Like - I got it working, but please note 

apiVersion: v1
kind: IOConfig
module: mewbot.io.desktop_notification
name: DesktopNotificationIO
uuid: 2
properties:
  token: "[should-not-be_needed]"

javajawa commented on examples/discord_to_desktop_notification.yaml:
```diff
@@ -0,0 +1,44 @@
+module: mewbot.io.desktop_notification
+name: DesktopNotificationIO
+uuid: 2
+properties:
+  token: "[should-not-be_needed]"
```
You might be able to make the properties problem go away with
```suggestion
properties: {}
```
(unless you have already tried that).

ajCameron replied:
I would have been prepared to swear I did ... but it would seem not!

javajawa commented on src/mewbot/io/desktop_notification.py:
```diff
@@ -0,0 +1,242 @@
+        # Note-to-self: Need to test to see if we are headless and fallback if we are
+
+        self._platform_str = sys.platform
+        self._logger.info(f"We are detected as running on {self._platform_str}")
```
I'm surprised that nothing is complaining about this not being lazily evaluated
```suggestion
        self._logger.info("We are detected as running on %s", self._platform_str)
```

(I'm not going to repeat the comment for each instance, but I should go look at why the linters aren't picking that up)

ajCameron replied:
Corrected. Also, after looking into it, not sure why the linters are not getting it.

javajawa commented on src/mewbot/io/desktop_notification.py:
```diff
@@ -0,0 +1,242 @@
+            self._detected_os = "windows"
+            self._do_windows_setup()
+
+        elif self._platform_str == "linux":
+            if self._platform_str == "linux2":
```
This is an impossible condition? (it can not be equal to both...surely)

exachixkitsune replied:
How tight is the setup.py's requirement on python 3.8+, will this likely ever occur, or is it a catch for edge cases?

ajCameron replied:
Derp! Indeed an impossible condition. Corrected.

It's highly unlikely to ever actually occur - indeed a catch for an edge case.

javajawa commented on src/mewbot/io/desktop_notification.py:
```diff
@@ -0,0 +1,242 @@
+            setattr(self, "_notify", self._linux_notify_send_method)
+
+        elif self._platform_str == "darwin":
+            self._detected_os = "macos"
+            self._enabled = False
```
No logging information about the OS in this branch.
Might be worth pulling out to a single log statement at the bottom with the detected OS and the enabled state?

ajCameron replied:
Agreed and done.

javajawa commented on src/mewbot/io/desktop_notification.py:
```diff
@@ -0,0 +1,242 @@
+class DesktopNotificationOutputEngine:
+    """
+    Contains the dubious and concerning logic required to make output work on all target systems.
+    """
```
The doc block should at least mention what this is and how it works ;)

```suggestion
    """
    Cross-Platform notification shim/engine.
    
    Works by patching the `_notify` function at init time to contain the correct implementation.
    """
```

ajCameron replied:
... I mean - that might be helpful...

javajawa commented on src/mewbot/io/desktop_notification.py:
```diff
@@ -0,0 +1,242 @@
+            return False
+        return self._notify(title, text)
+
+    def _notify(self, title: str, text: str) -> bool:
+        self._logger.warning(
+            f"{self._notify} should never be called directly - "
+            f"should have been live replaced with the true method - f{title} - f{text}"
+        )
+        return False
```
An alternative to this would be a dict of [detected OS; function to call] which might be easier to follow:
```suggestion
    def _notify(self, title: str, text: str) -> bool:
        # Probably don't do this inline...
        caller = {"windows": self._windows_toast_method, "linux": self._linux_notify_send_method}.get(self._detected_os)
        
        if not caller:
          self._logger.warning("No notification method supported for %s", self._detected_os)
          return False
          
        return caller(title, text)
```

exachixkitsune replied:
Does the setattr method appreciably reduce runtime?

javajawa replied:
`setattr` will be faster than what I've outlined here (as there won't essentially be a proxy to another method) and is almost certainly the fastest method for live matching something in standard python. On the flip side, if there's enough notifications that an extra layer of stack indirection makes a difference to real world performance, there's going to be so many notifications

ajCameron replied:
It is easier to read - correcting.

javajawa reviewed (commented):
Looks good codewise -- I've left a few comments here and there but nothing major.
Would probably be good to rewrite some of the commit messages to be more descriptive of the changes

exachixkitsune commented on setup.py:
```diff
@@ -25,6 +25,9 @@
         "Operating System :: OS Independent",
     ],
+    extras_require={
+        ':sys_platform == "win32"': ["win10toast~=0.9"],
```
Should this be added to requirements.txt?

ajCameron replied:
Did not realise that you could now - have corrected

javajawa replied:
I also had no idea you could do that. That's...going to be useful elsewhere

exachixkitsune commented on src/mewbot/io/desktop_notification.py:
```diff
@@ -0,0 +1,242 @@
+        # silent failure of desktop notifications is not desired. Noisy failure is.
+
+        self._enabled = False
+        if self._platform_str == "win32" or self._platform_str == "win64":
```
win64 doesn't appear to ever be a valid output from sys.platform
(source: https://docs.python.org/3.10/library/sys.html#sys.platform , not reporting win64 on my version)

ajCameron replied:
Removing.

ajCameron replied:
(What happens when you go off memory from a decade ago - instead of just reading the docs)

exachixkitsune reviewed (commented):
Had a look, all seems reasonable, though I've not actually tested the code.

ajCameron committed:
Implementing review suggestions

 - Changed the DesktopNotificationEngine to store the functions in a dict and retrieve from there
 - Moved the win10toast requirement logic fully into requirements.txt and removed it from setup.py

ajCameron committed:
Applied fixes to enable static logging

 - Static logging used in all cases
 - More helpful logging messages when trying to use notify2

ajCameron committed:
Minor black correction

And that's what happens when you manage to fat finger into the IDE while running the commit script :(

javajawa commented on examples/new 3.txt:
```diff
@@ -0,0 +1,2 @@
+    # ['summary_detail', 'published_parsed', 'links', 'title',
+    #  'summary', 'guidislink', 'title_detail', 'link', 'published', 'id']
```
Scratch file? :3

javajawa reviewed (commented):
Code approved. What looks like a scratch file got thrown in, should be easy enough to rebase out ;)

ajCameron committed:
Squashed scratch file

Which should have been on the rss dev branch.
